File: src/main/swift/Sources/swift/fr/jhelp/kotlinLigthTools/atomic/AtomicLong.swift


class AtomicLong {

    private let mutex: Mutex
    private var value: Long

    init(_ value: Long){
        self.mutex = Mutex()
        self.value = value
    }

    func get() -> Long
    {
        var value = Long(0)
        self.mutex.safeExecute { value = self.value }
        return value
    }

    func set(_ value: Long)
    {
        self.mutex.safeExecute { self.value = value }
    }

    func getAndSet(_ newValue: Long) -> Long
    {
        var value = Long(0)

        self.mutex.safeExecute {
            value = self.value
            self.value = newValue
        }

        return value
    }

    func compareAndSet(_ excepted: Long, _ newValue: Long) -> Bool
    {
        var done = false

        self.mutex.safeExecute {
            if self.value == excepted
            {
                self.value = newValue
                done = true
            }
        }

        return done
    }

    func incrementAndGet() -> Long
    {
        var value = Long(0)

        self.mutex.safeExecute {
            self.value += 1
            value = self.value
        }

        return value
    }

    func getAndIncrement() -> Long
    {
        var value = Long(0)

        self.mutex.safeExecute {
            value = self.value
            self.value += 1
        }

        return value
    }

    func decrementAndGet() -> Long
    {
        var value = Long(0)

        self.mutex.safeExecute {
            self.value -= 1
            value = self.value
        }

        return value
    }

    func getAndDecrement() -> Long
    {
        var value = Long(0)

        self.mutex.safeExecute {
            value = self.value
            self.value -= 1
        }

        return value
    }
}

File: src/main/kotlin/fr/jhelp/kotlinLigthTools/atomic/AtomicLong.kt
package fr.jhelp.kotlinLigthTools.atomic

import fr.jhelp.kotlinLight.Mutex

class AtomicLong
{
    private val mutex: Mutex
    private var value: Long

    constructor(value: Long)
    {
        this.mutex = Mutex()
        this.value = value
    }

    fun get(): Long
    {
        var value = 0L
        this.mutex.safeExecute { value = this.value }
        return value
    }

    fun set(value: Long)
    {
        this.mutex.safeExecute { this.value = value }
    }

    fun getAndSet(newValue: Long): Long
    {
        var value = 0L

        this.mutex.safeExecute {
            value = this.value
            this.value = newValue
        }

        return value
    }

    fun compareAndSet(excepted: Long, newValue: Long): Boolean
    {
        var done = false

        this.mutex.safeExecute {
            if (this.value == excepted)
            {
                this.value = newValue
                done = true
            }
        }

        return done
    }

    fun incrementAndGet(): Long
    {
        var value = 0L

        this.mutex.safeExecute {
            this.value++
            value = this.value
        }

        return value
    }

    fun getAndIncrement(): Long
    {
        var value = 0L

        this.mutex.safeExecute {
            value = this.value
            this.value++
        }

        return value
    }

    fun decrementAndGet(): Long
    {
        var value = 0L

        this.mutex.safeExecute {
            this.value--
            value = this.value
        }

        return value
    }

    fun getAndDecrement(): Long
    {
        var value = 0L

        this.mutex.safeExecute {
            value = this.value
            this.value--
        }

        return value
    }
}